### PR TITLE
net: l2: ieee802154: several minor improvements

### PR DESCRIPTION
**Pull request by ghost**
This PR introduces several minor improvements to the IEEE 802.15.4 L2 stack that I happened to stumble upon while developing for TSCH.

### CODE REVIEW

**Review by rlubos — DISMISSED** [review]
LGTM

File: subsys/net/l2/ieee802154/ieee802154.c [hunk 1]
```diff
@@ -453,12 +454,13 @@ static enum net_verdict ieee802154_recv(struct net_if *iface, struct net_pkt *pk
...
 #ifdef CONFIG_NET_6LO
 	verdict = ieee802154_6lo_decode_pkt(iface, pkt);
+#endif /* CONFIG_NET_6LO */
+
+	if (verdict == NET_CONTINUE) {
+		pkt_hexdump(RX_PKT_TITLE, pkt, true);
+	}
 
-	pkt_hexdump(RX_PKT_TITLE, pkt, true);
 	return verdict;
```
Comment (nandojve) [inline subsys/net/l2/ieee802154/ieee802154.c:463]: How you can guarantee that you have a valid verdict here if you dropped the default value?

File: subsys/net/l2/ieee802154/ieee802154.c [hunk 2]
```diff
@@ -387,8 +387,9 @@ static enum net_verdict ieee802154_recv(struct net_if *iface, struct net_pkt *pk
 
 	if (fs->fc.frame_type == IEEE802154_FRAME_TYPE_BEACON) {
 		verdict = ieee802154_handle_beacon(iface, &mpdu, net_pkt_ieee802154_lqi(pkt));
-		if (verdict == NET_OK) {
+		if (verdict == NET_CONTINUE) {
 			net_pkt_unref(pkt);
+			return NET_OK;
```
Comment (nandojve) [inline subsys/net/l2/ieee802154/ieee802154.c:392]: This changes (look only in the PR) makes me think that there is no central point to `unref` the frames.
Is there a clear way to `unref` a frame that can be used in general? I mean, each time a NET_OK will be returned (frame processed) the package should be `unref`. Is my assumption that correct?

File: subsys/net/lib/config/Kconfig [hunk 3]
```diff
@@ -171,7 +171,7 @@ config NET_CONFIG_IEEE802154_SECURITY_LEVEL
 	  1 authentication only with a 4 bytes length tag
 	  2 authentication only with a 8 bytes length tag
 	  3 authentication only with a 16 bytes length tag
-	  4 encryption only
+	  4 deprecated since IEEE 802.15.4-2015
```
Comment (nandojve) [inline subsys/net/lib/config/Kconfig:174]: Just to make this clear, there is a decision that Zephyr stack will not be compatible with previous IEEE 802.15.4 versions. Is that correct?

**Review by nandojve — CHANGES_REQUESTED** [review]

File: subsys/net/l2/ieee802154/ieee802154.c (hunk 2 above)
Comment (fgrandel) [reply subsys/net/l2/ieee802154/ieee802154.c:392]: Yes, exactly. That is a assumption valid throughout the net stack and that's also the very reason why I changed the semantics here - previously the internal methods would return NET_OK, although they did NOT unref the packages.

The correct way throughout the stack is:
* return NET_CONTINUE/NET_DROP: the upper layers are responsible for unref'ing the packet
* return NET_OK: upper layers will assume that the packet has been consumed and unref'ed

We still have a (well documented) exception in the driver API (ieee802154_handle_ack) - it is assumed that drivers will always unref the packet, even if this method returns NET_OK, that's for backwards compatibility and has a TODO attached to it.

File: subsys/net/lib/config/Kconfig (hunk 3 above)
Comment (fgrandel) [reply subsys/net/lib/config/Kconfig:174]: Yes, that option is a security risk and has therefore be deprecated and removed in recent versions of the standard.

**Review by jukkar — APPROVED** [review]

**Review by nandojve — APPROVED** [review]

**Pull request merged**